### PR TITLE
basic implementation of ui for notifications

### PR DESCRIPTION
**Pull request by samandmoore**
this is just the server side code and the view for notifications ui. i also did the js for marking an item as read. everything else is basic server-side logic for now.

we need to make it look better because there is 0 ui work on this, it's bare-bones

### CODE REVIEW

File: JabbR/Models/RepositoryExtensions.cs [hunk 1]
```diff
@@ -112,5 +112,15 @@ public static ChatUser VerifyUser(this IJabbrRepository repository, string userN
...
         }
+
+        public static IQueryable<Notification> Unread(this IQueryable<Notification> source)
+        {
+            return source.Where(n => !n.Read);
+        }
+
+        public static IQueryable<Notification> ByRoom(this IQueryable<Notification> source, string roomName)
+        {
+            return source.Where(n => n.Message.Room.Name == roomName);
```
Comment (davidfowl) [inline JabbR/Models/RepositoryExtensions.cs]: Nooooooo, inefficient. Add a method to the repository for this.

Comment (samandmoore) [inline JabbR/Models/RepositoryExtensions.cs]: EF profiler says we're okay!

```
SELECT TOP (20) [Project1].[Key]     AS [Key],
                [Project1].[Name]    AS [Name],
                [Project1].[Hash]    AS [Hash],
                [Project1].[Content] AS [Content],
                [Project1].[Name1]   AS [Name1],
                [Project1].[Read]    AS [Read]
FROM   (SELECT [Project1].[Key]                                    AS [Key],
               [Project1].[Read]                                   AS [Read],
               [Project1].[Content]                                AS [Content],
               [Project1].[When]                                   AS [When],
               [Project1].[Name]                                   AS [Name],
               [Project1].[Hash]                                   AS [Hash],
               [Project1].[Name1]                                  AS [Name1],
               row_number() OVER (ORDER BY [Project1].[When] DESC) AS [row_number]
        FROM   (SELECT [Extent1].[Key]     AS [Key],
                       [Extent1].[Read]    AS [Read],
                       [Extent2].[Content] AS [Content],
                       [Extent2].[When]    AS [When],
                       [Extent4].[Name]    AS [Name],
                       [Extent5].[Hash]    AS [Hash],
                       [Extent6].[Name]    AS [Name1]
                FROM   [dbo].[Notifications] AS [Extent1]
                       INNER JOIN [dbo].[ChatMessages] AS [Extent2]
                         ON [Extent1].[MessageKey] = [Extent2].[Key]
                       INNER JOIN [dbo].[ChatRooms] AS [Extent3]
                         ON [Extent2].[Room_Key] = [Extent3].[Key]
                       INNER JOIN [dbo].[ChatUsers] AS [Extent4]
                         ON [Extent1].[UserKey] = [Extent4].[Key]
                       LEFT OUTER JOIN [dbo].[ChatUsers] AS [Extent5]
                         ON [Extent2].[User_Key] = [Extent5].[Key]
                       LEFT OUTER JOIN [dbo].[ChatRooms] AS [Extent6]
                         ON [Extent2].[Room_Key] = [Extent6].[Key]
                WHERE  ([Extent1].[UserKey] = 2 /* @p__linq__0 */)
                       AND ([Extent3].[Name] = 'test' /* @p__linq__1 */)) AS [Project1]) AS [Project1]
WHERE  [Project1].[row_number] > 0
ORDER  BY [Project1].[When] DESC
```

File: JabbR/Migrations/201304082240509_AddRoomToNotification.cs [hunk 2]
```diff
@@ -0,0 +1,30 @@
...
+    using System;
+    using System.Data.Entity.Migrations;
+    
+    public partial class AddRoomToNotification : DbMigration
+    {
+        public override void Up()
+        {
+            AddColumn("dbo.Notifications", "RoomKey", c => c.Int(nullable: true));
+
+            Sql(@"UPDATE n
```
Comment (davidfowl) [inline JabbR/Migrations/201304082240509_AddRoomToNotification.cs:12]: Nice!

File: JabbR/Services/InMemoryRepository.cs [hunk 3]
```diff
@@ -138,6 +138,11 @@ public IQueryable<ChatRoom> GetAllowedRooms(ChatUser user)
                 .AsQueryable();
         }
 
+        public IQueryable<Notification> GetNotificationsByUser(ChatUser user)
+        {
+            return _notifications.Where(n => n.UserKey == user.Key).AsQueryable();
```
Comment (davidfowl) [inline JabbR/Services/InMemoryRepository.cs:143]: You don't need AsQueryable

Comment (davidfowl) [inline JabbR/Services/InMemoryRepository.cs:143]: Oops this is in memory.

File: JabbR/Views/Notifications/index.cshtml [hunk 4]
```diff
@@ -0,0 +1,149 @@
...
+                    {
+                        <p class="alert alert-block alert-info">No notifications!</p>
+                    }
+                </div>
+            </div>
+        </div>
+    </section>
+    <script src="@Url.Content("~/Scripts/jquery-1.9.0.min.js")"></script>
+    <script src="@Url.Content("~/Scripts/bootstrap.min.js")"></script>
+    <script type="text/javascript">
```
Comment (davidfowl) [inline JabbR/Views/Notifications/index.cshtml]: Separate file for this js,

File: JabbR/Views/Home/index.cshtml [hunk 5]
```diff
@@ -255,6 +255,7 @@
       </div>
         <div id="preferences" class="btn-group account pull-right">
             <a class="btn help" aria-haspopup="true" title="display help"><i class="icon-question-sign"></i></a>
+            <a class="btn" title="notifications" href="@Url.Content("~/notifications")" target="_blank"><i class="icon-comments"></i></a>
```
Comment (davidfowl) [inline JabbR/Views/Home/index.cshtml:258]: This needs to be live and show unread in the future.

File: JabbR/Views/Notifications/index.cshtml (hunk 4 above)
Comment (samandmoore) [inline JabbR/Views/Notifications/index.cshtml]: Will do tomorrow.

File: JabbR/Views/Home/index.cshtml (hunk 5 above)
Comment (samandmoore) [inline JabbR/Views/Home/index.cshtml:258]: Yeah we can add a method on the hub for keeping it in sync.

File: JabbR/Views/Notifications/index.cshtml (hunk 4 above)
Comment (samandmoore) [inline JabbR/Views/Notifications/index.cshtml]: I need Henrik to figure out how to make the icon look cool with the count
though

Comment (hnrkndrssn) [inline JabbR/Views/Notifications/index.cshtml]: I'll get on this once I can clean up the room loading notification :smile:

**Comment by davidfowl** [comment]
@samandmoore looks good. Merge it in when you're ready I wanna start playing with it!

**Pull request merged**